File: bigdime-data-handlers/src/main/java/io/bigdime/handler/webhdfs/WebHDFSInputDescriptor.java
package io.bigdime.handler.webhdfs;

import java.nio.channels.ReadableByteChannel;
import java.util.List;

import io.bigdime.core.InputDescriptor;
import io.bigdime.core.commons.StringHelper;
import io.bigdime.libs.hdfs.FileStatus;

class WebHDFSInputDescriptor implements InputDescriptor<String> {
	private static final String INPUT_DESCRIPTOR_PREFIX = "handlerClass:io.bigdime.handler.webhdfs.WebHDFSReaderHandler,webhdfsPath:";

	private String fullDescriptor;// complete filename, with prefix
	private String webhdfsPath;// complete filename

	private FileStatus currentFileStatus;
	private String currentFilePath;
	// private long fileLength = -1;
	private ReadableByteChannel fileChannel;

	@Override
	public String getNext(List<String> availableInputDescriptors, String lastInputDescriptor) {
		int indexOfLastInput = availableInputDescriptors.indexOf(lastInputDescriptor);
		if (availableInputDescriptors.size() > indexOfLastInput + 1) {
			return availableInputDescriptors.get(indexOfLastInput + 1);
		}
		return null;

	}

	@Override
	public void parseDescriptor(String inputDescriptor) {

		if (StringHelper.isBlank(inputDescriptor)) {
			throw new IllegalArgumentException("descriptor can't be null or empty");
		}

		if (inputDescriptor.indexOf(INPUT_DESCRIPTOR_PREFIX) < 0) {
			throw new IllegalArgumentException("descriptor must contain prefix:" + INPUT_DESCRIPTOR_PREFIX);

		}
		fullDescriptor = inputDescriptor;
		webhdfsPath = inputDescriptor.substring(INPUT_DESCRIPTOR_PREFIX.length());
		currentFilePath = webhdfsPath;
	}

	public String getWebhdfsPath() {
		return webhdfsPath;
	}

	public void setWebhdfsPath(String webhdfsPath) {
		this.webhdfsPath = webhdfsPath;
	}

	public String getFullDescriptor() {
		return fullDescriptor;
	}

	public String getDescriptorPrefix() {
		return INPUT_DESCRIPTOR_PREFIX;
	}

	public String createFullDescriptor(String webhdfsPath) {
		return INPUT_DESCRIPTOR_PREFIX + webhdfsPath;
	}

	public FileStatus getCurrentFileStatus() {
		return currentFileStatus;
	}

	public void setCurrentFileStatus(FileStatus currentFileStatus) {
		this.currentFileStatus = currentFileStatus;
	}

	public String getCurrentFilePath() {
		return currentFilePath;
	}

	public void setCurrentFilePath(String currentFilePath) {
		this.currentFilePath = currentFilePath;
	}

	public long getFileLength() {
		return getCurrentFileStatus().getLength();
	}

	public ReadableByteChannel getFileChannel() {
		return fileChannel;
	}

	public void setFileChannel(ReadableByteChannel fileChannel) {
		this.fileChannel = fileChannel;
	}
}

File: bigdime-data-handlers/src/main/scala/io/bigdime/handler/webhdfs/WebHDFSInputDescriptor.scala
package io.bigdime.handler.webhdfs

import java.nio.channels.ReadableByteChannel

import io.bigdime.core.InputDescriptor
import io.bigdime.core.commons.StringHelper
import io.bigdime.libs.hdfs.FileStatus

/**
  * Created by neejain on 12/11/16.
  */
object WebHDFSInputDescriptor {
  //  private val INPUT_DESCRIPTOR_PREFIX = "handlerClass:io.bigdime.handler.webhdfs.WebhdfsReaderAndSink,webhdfsPath:"
}

class WebHDFSInputDescriptor(inputDescriptorPrefix: String = "handlerClass:io.bigdime.handler.webhdfs.WebHDFSReaderHandler,webhdfsPath:") extends InputDescriptor[String] {
  private var fullDescriptor: String = null
  // complete filename, with prefix
  private var webhdfsPath: String = null
  // complete filename
  private var currentFileStatus: FileStatus = null
  private var currentFilePath: String = null
  // private long fileLength = -1;
  private var fileChannel: ReadableByteChannel = null


  def getNext(availableInputDescriptors: java.util.List[String], lastInputDescriptor: String): String = {
    val indexOfLastInput: Int = availableInputDescriptors.indexOf(lastInputDescriptor)
    if (availableInputDescriptors.size > indexOfLastInput + 1) {
      return availableInputDescriptors.get(indexOfLastInput + 1)
    }
    return null
  }

  def parseDescriptor(inputDescriptor: String) {
    if (StringHelper.isBlank(inputDescriptor)) {
      throw new IllegalArgumentException("descriptor can't be null or empty")
    }
    if (inputDescriptor.indexOf(inputDescriptorPrefix) < 0) {
      throw new IllegalArgumentException("descriptor must contain prefix:" + inputDescriptorPrefix)
    }
    fullDescriptor = inputDescriptor
    webhdfsPath = inputDescriptor.substring(inputDescriptorPrefix.length)
    currentFilePath = webhdfsPath
  }

  def getWebhdfsPath: String = {
    return webhdfsPath
  }

  def setWebhdfsPath(webhdfsPath: String) {
    this.webhdfsPath = webhdfsPath
  }

  def getFullDescriptor: String = {
    return fullDescriptor
  }

  def getDescriptorPrefix: String = {
    return inputDescriptorPrefix
  }

  def createFullDescriptor(webhdfsPath: String): String = {
    return inputDescriptorPrefix + webhdfsPath
  }

  def getCurrentFileStatus: FileStatus = {
    return currentFileStatus
  }

  def setCurrentFileStatus(currentFileStatus: FileStatus) {
    this.currentFileStatus = currentFileStatus
  }

  def getCurrentFilePath: String = {
    return currentFilePath
  }

  def setCurrentFilePath(currentFilePath: String) {
    this.currentFilePath = currentFilePath
  }

  def getFileLength: Long = {
    return getCurrentFileStatus.length
  }

  def getFileChannel: ReadableByteChannel = {
    return fileChannel
  }

  def setFileChannel(fileChannel: ReadableByteChannel) {
    this.fileChannel = fileChannel
  }

  def this() {
    this("handlerClass:io.bigdime.handler.webhdfs.WebHDFSReaderHandler,webhdfsPath:")
  }
}